### DISCUSSION TITLE
How to execture raw SQL queries in `SchemaGenerator`?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mikro-orm/mikro-orm#3979] [Q&A] [answered]
I need to create some views (see #672) as part during schema creation (I created one based on [this example](https://mikro-orm.io/docs/schema-generator#using-schemagenerator-programmatically)), therefore I wanted to add some `orm.execute()` or `orm.raw()` commands.

However, [`const orm = await MikroORM.init()`](https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/MikroORM.ts#L14) does not provide `execute()` nor `raw()` methods.

Is it somehow possible to execute raw SQL queries in the context of `SchemaGenerator`?

I use `mikro-orm` version `5.6.8` with PostgreSQL in `node@19.5.0` with `typescript@4.9.4`.

**Comment by tukusejssirs** [comment]
I’ve found it: `await orm.em.getConnection().execute(`${sql}`)`

**Comment by B4nan** [comment, marked as answer]
Your question was "How to execture raw SQL queries in `SchemaGenerator`?", and `SchemaGenerator` has `execute` method.

```ts
await orm.schema.execute('...');
```

**Reply by tukusejssirs** [reply]
Okay, I might misunderstand the MikroORM nomenclature, but you used the term _`SchemaGenerator`_ in the [docs](https://mikro-orm.io/docs/schema-generator#using-schemagenerator-programmatically) right about the script to create DB schema programmatically. I was referring to this.

By _raw SQL queries_ I meant a way to run any SQL command supported by a particular DB. For this purpose, both `orm.em.getConnection().execute()` and `orm.schema.execute()` work as expected. What is the difference between the too (if any)?

**Reply by B4nan** [reply]
`schema.execute()` supports passing full schema definition, one command per line, the connection method requires only a single statement (unless the driver supports multi-statements automatically). On the other hand, the connection method allows you to pass in knex QB instance too.
